cklib/cklib/x509.py

```diff
@@ -0,0 +1,228 @@
+import os
+from ipaddress import (
+    ip_address,
+    ip_network,
+    IPv4Address,
+    IPv6Address,
+    IPv4Network,
+    IPv6Network,
+)
+from datetime import datetime, timedelta, timezone
+from cryptography import x509
+from cryptography.x509.oid import NameOID
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateKey
+from cryptography.x509.base import Certificate, CertificateSigningRequest
+from typing import List, Optional, Tuple, Union
+
+
+def gen_rsa_key(key_size: int = 2048) -> RSAPrivateKey:
+    return rsa.generate_private_key(
+        public_exponent=65537, key_size=key_size, backend=default_backend()
+    )
+
+
+def bootstrap_ca(
+    days_valid: int = 3650,
+    common_name: str = "Cloudkeeper Certificate Authority",
+    organization_name: str = "Some Engineering Inc.",
+    locality_name: str = "San Francisco",
+    state_or_province_name: str = "California",
+    country_name: str = "US",
+) -> Tuple[RSAPrivateKey, Certificate]:
+    ca_key = gen_rsa_key()
+    subject = issuer = x509.Name(
+        [
+            x509.NameAttribute(NameOID.COUNTRY_NAME, country_name),
+            x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, state_or_province_name),
+            x509.NameAttribute(NameOID.LOCALITY_NAME, locality_name),
+            x509.NameAttribute(NameOID.ORGANIZATION_NAME, organization_name),
+            x509.NameAttribute(NameOID.COMMON_NAME, common_name),
+        ]
+    )
+    ca_cert = (
+        x509.CertificateBuilder()
+        .subject_name(subject)
+        .issuer_name(issuer)
+        .public_key(ca_key.public_key())
+        .serial_number(x509.random_serial_number())
+        .not_valid_before(datetime.utcnow().replace(tzinfo=timezone.utc))
+        .not_valid_after(
+            datetime.utcnow().replace(tzinfo=timezone.utc) + timedelta(days=days_valid)
+        )
+        .add_extension(x509.BasicConstraints(ca=True, path_length=None), critical=True)
+        .add_extension(
+            x509.KeyUsage(
+                digital_signature=False,
+                key_encipherment=False,
+                key_cert_sign=True,
+                key_agreement=False,
+                content_commitment=False,
+                data_encipherment=False,
+                crl_sign=True,
+                encipher_only=False,
+                decipher_only=False,
```

Comment:
These are x509 capabilities universal. They define what the certificate is allowed and not allowed to do. Since this is the CA cert it's only allowed to sign other certs as well as the certificate revocation list. Further down where we sign the CSR you'll see that those certs are only allowed to encrypt traffic but not sign other certificates. If we wanted to have E.g. a root CA and then intermediate CAs we could issue certs with the capability to sign others and create a longer cert path. That way we'd have a train of trust.

Reply:
Sorry - I mean it would be great to have that comment in the code.